File: bms/go/geo/geo_s2_Cap_Centroid/geo_s2_Cap_Centroid.go
package test

import (
	"math"
)

func (c geo_s2_Cap) Centroid() geo_s2_Point {
	// From symmetry, the centroid of the cap must be somewhere on the line
	// from the origin to the center of the cap on the surface of the sphere.
	// When a sphere is divided into slices of constant thickness by a set of
	// parallel planes, all slices have the same surface area. This implies
	// that the radial component of the centroid is simply the midpoint of the
	// range of radial distances spanned by the cap. That is easily computed
	// from the cap height.
	if c.IsEmpty() {
		return geo_s2_Point{}
	}
	r := 1 - 0.5*c.Height()
	return geo_s2_Point{c.center.Mul(r * c.Area())}
}

func (c geo_s2_Cap) IsEmpty() bool {
	return c.radius < 0
}

func (c geo_s2_Cap) Height() float64 {
	return float64(0.5 * c.radius)
}

func (c geo_s2_Cap) Area() float64 {
	return 2.0 * math.Pi * math.Max(0, c.Height())
}

func (v geo_r3_Vector) Mul(m float64) geo_r3_Vector { return geo_r3_Vector{m * v.X, m * v.Y, m * v.Z} }

type geo_s2_Cap struct {
	center geo_s2_Point
	radius geo_s1_ChordAngle
}

type geo_s2_Point struct {
	geo_r3_Vector
}

type geo_r3_Vector struct {
	X, Y, Z float64
}

type geo_s1_ChordAngle float64

File: translations_only/fail/go/geo/geo_s2_Cap_Centroid/mistral/CAPR/geo_s2_Cap_Centroid.rs

use std::f64;

struct GeoR3Vector {
    x: f64,
    y: f64,
    z: f64,
}

struct GeoS1ChordAngle(f64);

struct GeoS2Point {
    geo_r3_vector: GeoR3Vector,
}

struct GeoS2Cap {
    center: GeoS2Point,
    radius: GeoS1ChordAngle,
}

impl GeoS2Cap {
    fn centroid(&self) -> GeoS2Point {
        if self.is_empty() {
            return GeoS2Point {
                geo_r3_vector: GeoR3Vector { x: 0.0, y: 0.0, z: 0.0 },
            };
        }
        let r = 1.0 - 0.5 * self.height();
        GeoS2Point {
            geo_r3_vector: GeoR3Vector {
                x: self.center.geo_r3_vector.x * r * self.area(),
                y: self.center.geo_r3_vector.y * r * self.area(),
                z: self.center.geo_r3_vector.z * r * self.area(),
            },
        }
    }

    fn is_empty(&self) -> bool {
        self.radius.0 < 0.0
    }

    fn height(&self) -> f64 {
        0.5 * self.radius.0
    }

    fn area(&self) -> f64 {
        2.0 * f64::consts::PI * f64::max(0.0, self.height())
    }
}

impl GeoR3Vector {
    fn mul(&self, m: f64) -> GeoR3Vector {
        GeoR3Vector {
            x: m * self.x,
            y: m * self.y,
            z: m * self.z,
        }
    }
}
